pkg/fileutils/read.go

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2018 The original author or authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package fileutils
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+// Read reads the contents of the specified file. If the file is a relative path, it is relative to base.
+// Either file or base may be URLs.
+func Read(file string, base string) ([]byte, error) {
+	absoluteFile, err := absFile(file, base)
+	if err != nil {
+		return nil, err
+	}
+	return readAbsFile(absoluteFile)
+}
+
+func absFile(file string, base string) (string, error) {
+	u, err := url.Parse(file)
+	if err != nil {
+		return "", err
+	}
+	if u.IsAbs() {
+		return u.String(), nil
+	}
+
+	if filepath.IsAbs(file) {
+		return file, nil
+	}
+
+	b, err := url.Parse(base)
+	if err != nil {
+		return "", err
+	}
+	if b.IsAbs() {
+		b.Path = path.Join(b.Path, file)
+		return b.String(), nil
+	}
+
+	if filepath.IsAbs(base) {
```

Comment:
Can this be true while line 58 returned false?

Reply:
No it can't. This was defensive programming in case `fileutils` grows and someone accidentally calls `readAbsFile` with a relative path.

Reply:
I'm not sure I follow. This is `absFile`, and it doesn't interact with `readAbsFile` (`Read` does).

Reply:
Apologies for answering a question you weren't asking!

`IsAbs` from the `url` package returns `true` if and only if the scheme is non-empty. So `IsAbs` will return `false` if we pass a non-URL path, regardless of whether that path is absolute or relative.